File: src/main/java/ru/hanoi/Hanoi4.java
package ru.hanoi;

import org.junit.Test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;

public class Hanoi4 {
    @Test public void findMovesToSolveTowersOfHanoiProblem() {
        assertThat(solveHanoi(1), equalTo(asList(move(0, -1))));
        assertThat(solveHanoi(2), equalTo(asList(move(0, 1), move(1, -1), move(0, 1))));
        assertThat(solveHanoi(3), equalTo(asList(
                move(0, -1), move(1, 1), move(0, -1),
                move(2, -1),
                move(0, -1), move(1, 1), move(0, -1)
        )));
    }

    private List<Move> solveHanoi(int i) {
        return solveHanoi(i, -1);
    }

    private List<Move> solveHanoi(int i, int direction) {
        if (i == 0) return Collections.emptyList();
        else {
            List<Move> result = new ArrayList<>();
            result.addAll(solveHanoi(i - 1, -direction));
            result.add(move(i - 1, direction));
            result.addAll(solveHanoi(i - 1, -direction));
            return result;
        }
    }

    private static Move move(int plate, int shift) {
        return new Move(plate, shift);
    }

    private static class Move {
        final int plate;
        final int shift;

        private Move(int plate, int shift) {
            this.plate = plate;
            this.shift = shift;
        }

        @Override public String toString() {
            return "(" + plate + "," + shift + ")";
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Move move = (Move) o;

            if (plate != move.plate) return false;
            if (shift != move.shift) return false;

            return true;
        }

        @Override
        public int hashCode() {
            int result = plate;
            result = 31 * result + shift;
            return result;
        }
    }
}

File: scala/src/ru/hanoi/Hanoi4.scala
package ru.hanoi

import org.scalatest.junit.AssertionsForJUnit
import org.junit.Test
;

/*
 * User: dima
 * Date: 9/4/11
 * Time: 2:21 PM
 */
class Hanoi4 extends AssertionsForJUnit {
  @Test def shouldSolveHanoi() {
    assert(solveHanoi(1) === List(Move(1, 1)))
    assert(solveHanoi(2) === List(Move(1, -1), Move(2, 1), Move(1, -1)))
    assert(solveHanoi(3) === List(Move(1, 1), Move(2, -1), Move(1, 1), Move(3, 1), Move(1, 1), Move(2, -1), Move(1, 1)))
  }

  case class Move(disk: Int, direction: Int)

  def solveHanoi(hanoiSize: Int, direction: Int = 1): List[Move] = {
    if (hanoiSize == 0) // forgot to add this condition
      List()
    else
      solveHanoi(hanoiSize - 1, -direction) :::
        List(Move(hanoiSize, direction)) ::: // tried to append to just one element not a List
        solveHanoi(hanoiSize - 1, -direction)
  }
}